va/va.go

```diff
@@ -639,23 +640,35 @@ func (va *ValidationAuthorityImpl) validateChallenge(ctx context.Context, identi
 	return nil, probs.Malformed("invalid challenge type %s", challenge.Type)
 }
 
-func (va *ValidationAuthorityImpl) performRemoteValidation(ctx context.Context, domain string, challenge core.Challenge, authz core.Authorization, result chan *probs.ProblemDetails) {
-	s := va.clk.Now()
-	errors := make(chan error, len(va.remoteVAs))
+// performRemoteValidation calls `PerformValidation` for each of the configured
+// remoteVAs in a random order. The provided `results` chan should have an equal
+// size to the number of remote VAs. The validations will be peformed in
+// separate go-routines. If the result `error` from a remote
+// `PerformValidation` RPC is nil or a nil `ProblemDetails` instance it is
+// written directly to the `results` chan. If the err is a cancelled error it is
+// treated as a nil error. Otherwise the error/problem is written to the results
+// channel as-is.
+func (va *ValidationAuthorityImpl) performRemoteValidation(
+	ctx context.Context,
+	domain string,
+	challenge core.Challenge,
+	authz core.Authorization,
+	results chan error) {
+	va.log.Infof("Performing remote validations")
 	for _, i := range rand.Perm(len(va.remoteVAs)) {
 		remoteVA := va.remoteVAs[i]
-		go func(rva RemoteVA) {
+		go func(rva RemoteVA, index int) {
 			_, err := rva.PerformValidation(ctx, domain, challenge, authz)
 			if err != nil {
 				// returned error can be a nil *probs.ProblemDetails which breaks the
 				// err != nil check so do a slightly more complicated unwrap check to
 				// make sure we don't choke on that.
-				if p, ok := err.(*probs.ProblemDetails); ok || p != nil {
+				if p, ok := err.(*probs.ProblemDetails); ok && p != (*probs.ProblemDetails)(nil) {
```

Comment:
I'll have to go back and rebuild the context for this change. It was required otherwise `nil` problem details weren't being handled correctly. I believe when we tried to fix this in the past it wasn't done correctly.

Reply:
`ok || p != nil` is just plain wrong and I had to change it in order to get this overall PR to work correctly.

We know that `PerformValidation` can return a `*probs.ProblemDetails(nil)` instance. In that case `ok` is true because `*probs.ProblemDetails(nil)` is a `*probs.ProblemDetails`. The first clause will short-circuit because `ok == true` so `ok || p != nil` returns true. That ends up treating the result a a normal negative validation result that should be logged when it isn't one and the outer `err` isn't cleared like it should be by the next `if` clause.

Using `ok && p != (*probs.ProblemDetails)(nil)` does the expected: `ok` is true because it is a problem details, but its `*probs.ProblemDetails(nil)` so the second part of the clause is false and the expression is false. The _next_ if clause handles this case as desired and we don't treat it as a failure or a negative result.

Reply:
That makes sense, thanks for explaining.